enterprise_catalog/apps/api/v1/serializers.py

```
@@ -26,6 +26,48 @@
 logger = logging.getLogger(__name__)
 
 
+def find_and_modify_catalog_query(content_filter, catalog_query_uuid=None):
+    """
+    This method aims to make sure UUID and content_filter in the catalog service
+    match what Django Admin/passed in parameters have. We take the parameters as source of truth,
+    but do not want to duplicate either UUID or content filter.
+
+    Arguments:
+        content_filter(dict): filter used to pick which courses are retrieved
+        catalog_query_uuid(UUID/str): query uuid generated from LMS Django Admin.
+            - If not provided, we should be receiving a "direct" content filter.
+    Returns:
+        a CatalogQuery object.
+    """
+    if catalog_query_uuid:
+        uuid_cg = CatalogQuery.objects.filter(uuid=catalog_query_uuid).first()
+        if uuid_cg:
+            uuid_cg.content_filter = content_filter
+            uuid_cg.content_filter_hash = get_content_filter_hash(content_filter)
+            uuid_cg.save()
```

So for our edge case, ie we get sent a "new" content filter that is actually a duplicate, how do we want to handle the inevitable IntegrityError thrown here?
Because currently it would throw a 500

---

Something like that should return a 422 response code, IMO.  You could raise a `ValidationError` (or a customer subclass thereof) in here and let that bubble up to the view, then handle it in the view.

---

Interesting. I just noticed for the EnterpriseCatalog `create` here we're just raising the IntegrityError and don't appear to be catching that error in the views either.

To handle this in the Views what would I wrap in the try/except? I see some instances where we override the HTTP methods (such as Post/Patch/get/etc.), but I'm not sure what logic would actually exist in the view function to wrap and how it's affected by us flipping which serializer we use based on the http command.

---

Any further thoughts here @iloveagent57 ?

---

Ah, sorry I mislead you into the view; this is one of the cases where the convenience of DRF shines.  `ValidationError` nicely lets you specify a status code that's used in the response: https://www.django-rest-framework.org/api-guide/exceptions/#validationerror
So for this duplicate case, you can do something like ```raise serializers.ValidationError({'catalog_query': 'UUID is not unique'}, code=status.HTTP_422_UNPROCESSABLE_ENTITY)```

---

Mmmk so that is being caught and the error message is coming through great. However, 422 is getting reduced to just a 400. Not sure why or if that's a big issue really.

---

upon some searching, it looks like the "is_valid" run on serializers may return its own ValidationError which always returns 400. I'd have to override the `is_valid` function to get a custom error message. However, I may do a little more digging, because why have a code parameter if it doesn't use it properly.

---

Ah, ok.  I wouldn't worry about 422 vs. 400, as long as the detail is coming through with a 4xx code.